l5kit/l5kit/simulation/dataset.py

```diff
@@ -0,0 +1,217 @@
+                 disable_new_agents: bool = False, distance_th_far: float = 15,
+                 distance_th_close: float = 10) -> None:
+        """This class allows to:
+        - rasterise the same frame across multiple scenes for ego;
+        - rasterise the same frame across multiple scenes for multiple agents;
+        - filter agents based on distance to ego;
+        - set ego in future frames;
+        - set agents in future frames;
+
+        .. note:: only vehicles (car label) are picked as agents
+
+        :param dataset: the original ego dataset
+        :param scene_indices: the indices of the scenes to take
+        :param start_frame: the unroll start frame, use only if disable_new_agents is True
+        :param disable_new_agents: if to disable new agents coming in after start_frame, defaults to False
+        :param distance_th_far: the distance threshold for new agents,
+            An agent will be grabbed if it's closer than this value to ego, defaults to 15
+        :param distance_th_close: the distance threshold for already tracked agents,
+            An agent will be grabbed if it's closer than this value to ego, defaults to 10
+        """
+        self.scene_indices = scene_indices
+        self.filter_agents_thr = dataset.cfg["raster_params"]["filter_agents_threshold"]
+
+        self.scene_dataset_batch: Dict[int, EgoDataset] = {}  # dicts preserve insertion order
+        for scene_idx in self.scene_indices:
+            scene_dataset = dataset.get_scene_dataset(scene_idx)
+            self.scene_dataset_batch[scene_idx] = scene_dataset
+
+        # keep track of original dataset
+        self.recorded_scene_dataset_batch = deepcopy(self.scene_dataset_batch)
+
+        # buffer used to keep track of tracked agents during unroll as tuples of scene_idx, agent_idx
+        self._agents_tracked: Set[Tuple[int, int]] = set()
+
+        self.distance_th_far = distance_th_far
+        self.distance_th_close = distance_th_close
+        self.disable_new_agents = disable_new_agents
+
+        if disable_new_agents:
+            # we disable all agents that wouldn't be picked at start_frame
+            for scene_index in scene_indices:
+                dataset_zarr = self.scene_dataset_batch[scene_index].dataset
+                frame = dataset_zarr.frames[start_frame]
+                ego_pos = frame["ego_translation"][:2]
+                agents = dataset_zarr.agents
+                frame_agents = filter_agents_by_frames(frame, agents)[0]
+                frame_agents = self._filter_agents(0, frame_agents, ego_pos)
+                disable_agents(dataset_zarr, allowlist=frame_agents["track_id"])
+
+    def __len__(self) -> int:
+        """
+        Return the minimum number of frames across scenes
+
+        :return: the number of frames
+        """
+        num_frame_batch = [len(scene_dt.dataset.frames) for scene_dt in self.scene_dataset_batch.values()]
+        return min(num_frame_batch)
```

Comment:
Is this going to change or be constant ? If it doesn't change, we can move this to the constructor and just return it here to avoid this counting at every `len()` call.